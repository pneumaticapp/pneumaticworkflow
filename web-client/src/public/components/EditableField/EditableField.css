.edit {
  display: flex;
  align-items: center;
}

.edit-field {
  margin: -5px;
  padding-right: 5px;
  display: block;
  width: calc(100% - 30px);
  height: 32px;
  border: 5px solid transparent;
}

.edit--hidden-icon .edit-field {
  padding: 0;
  width: 100%;
}

.edit--editable {
  .edit-icon {
    font-weight: 900;
    color: #e4af31;
  }
}

.edit-icon {
  margin-left: 30px;
  cursor: pointer;
}
