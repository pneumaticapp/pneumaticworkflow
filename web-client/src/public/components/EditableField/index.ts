/* eslint-disable */
/* prettier-ignore */
export * from './EditableField';
