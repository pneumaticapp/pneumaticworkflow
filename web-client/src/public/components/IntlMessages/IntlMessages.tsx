/* eslint-disable */
/* prettier-ignore */
import * as React from 'react';
import { FormattedMessage, injectIntl, IntlShape } from 'react-intl';

interface IInjectedOptions {
  id: string;
  intl: IntlShape;
}

type TInjectMessageProps = React.ComponentProps<typeof FormattedMessage> & IInjectedOptions;

const InjectMassage = (props: TInjectMessageProps) => <FormattedMessage {...props} tagName={props.tagName || 'span'}/>;

export const IntlMessages = injectIntl<'intl', TInjectMessageProps>(InjectMassage, {forwardRef: true});
