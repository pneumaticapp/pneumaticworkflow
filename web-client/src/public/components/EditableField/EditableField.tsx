/* eslint-disable */
/* prettier-ignore */
import * as React from 'react';
import * as classnames from 'classnames';
import styles from './EditableField.css';
import { onEnterPressed } from '../../utils/handlers';
import { Tooltip } from 'reactstrap';
import { IntlMessages } from '../IntlMessages';
import { sanitizeText } from '../../utils/strings';

export interface IEditableFieldProps {
  value: string;
  id: string;
  hiddenIcon?: boolean;
  changeValue(value: string): void;
  fieldClassName?: string;
  validate?(value: string): string;
}

export interface IEditableFieldState {
  isEditable: boolean;
  touched: boolean;
}

export class EditableField extends React.Component<IEditableFieldProps, IEditableFieldState> {
  public state = {
    isEditable: false,
    touched: false,
  };

  public shouldComponentUpdate(nextProps: IEditableFieldProps, nextState: IEditableFieldState): boolean {
    return !this.isTextEqual(nextProps.value, this.textContent) ||
      nextState.isEditable !== this.state.isEditable;
  }

  public fieldRef = React.createRef<HTMLSpanElement>();
  public render() {
    const {
      id,
      hiddenIcon,
      fieldClassName,
      value,
    } = this.props;
    const { isEditable } = this.state;
    const error = this.getError();

    return (
      <span
        id={id}
        className={classnames(
          styles['edit'],
          {[styles['edit--editable']]: isEditable, [styles['edit--hidden-icon']]: hiddenIcon},
        )}
      >
        <span
          className={classnames(styles['edit-field'], fieldClassName)}
          ref={this.fieldRef}
          contentEditable={isEditable}
          onBlur={this.blockEditable}
          onInput={this.handleChange}
          onKeyDown={onEnterPressed(this.blockEditable)}
          onClick={hiddenIcon ? this.onEditClick : undefined}
          suppressContentEditableWarning
        >
          {sanitizeText(value)}
        </span>
        {!hiddenIcon &&
          <i
            className={classnames('simple-icon-pencil', styles['edit-icon'])}
            onClick={this.onEditClick}
          />
        }
        {error &&
          <Tooltip
            placement="left"
            isOpen
            target={id}
          >
            <IntlMessages id={error}/>
          </Tooltip>
        }
      </span>
    );
  }

  private handleChange = (e: React.FormEvent) => {
    this.props.changeValue(sanitizeText(e.currentTarget.textContent));
  };

  private isTextEqual = (text1: string | null, text2: string | null) => {
    return sanitizeText(text1) === sanitizeText(text2);
  };

  private get textContent() {
    return this.fieldRef.current && this.fieldRef.current.textContent || '';
  }

  private getError = () => {
    const { validate } = this.props;
    const { touched } = this.state;
    const value = this.textContent;
    if (!validate || !touched) {
      return '';
    }

    return validate(value) || '';
  };

  private onEditClick = () => {
    if (!this.state.isEditable) {
      this.setState({isEditable: true}, this.handleCaret);
    }
  };

  private blockEditable = () => this.setState({isEditable: false, touched: true});

  private handleCaret = () => {
    const {current: element} = this.fieldRef;
    if (!element) {
      return;
    }
    element.focus();
    if (window.getSelection && document.createRange) {
      let range = document.createRange();
      range.selectNodeContents(element);
      range.collapse(false);
      let sel = window.getSelection();
      if (sel) {
        sel.removeAllRanges();
        sel.addRange(range);
      }
    }
  };
}
