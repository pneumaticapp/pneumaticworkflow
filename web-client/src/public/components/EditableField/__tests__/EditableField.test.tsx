import * as React from 'react';
import { shallow } from 'enzyme';

import { EditableField } from '../EditableField';
import { IntlMessages } from '../../IntlMessages';

const mockProps = {
  value: 'test',
  id: 'test-id',
  changeValue: jest.fn(),
};

describe('EditableField', () => {
  it('корректно рендерится', () => {
    const wrapper = shallow(<EditableField {...mockProps}/>);

    expect(wrapper).toMatchSnapshot();
  });
  it('рендерит ошибку, если значение не проходит валидацию', () => {
    const errorMsg = 'Length should be minimum five';
    const validate = (value: string) => value.length >= 5 ? '' : errorMsg;
    const wrapper = shallow(<EditableField {...mockProps} validate={validate} hiddenIcon />);
    wrapper.setState({touched: true});

    expect(wrapper.find(IntlMessages).prop('id')).toEqual(errorMsg);
  });
  it('если не рендерит иконку, то клик на элементе позволяет редактировать', () => {
    const wrapper = shallow(<EditableField {...mockProps} hiddenIcon />);

    wrapper.find('span > span').simulate('click');

    expect(wrapper.state()).toEqual(expect.objectContaining({isEditable: true}));
  });
  it('при изменении вызывает changeValue из пропсов', () => {
    const textContent = 'edit result';
    const wrapper = shallow(<EditableField {...mockProps} hiddenIcon />);

    wrapper.find('span > span').simulate('input', {currentTarget: {textContent}});

    expect(mockProps.changeValue).toHaveBeenCalledWith(textContent);
  });
  it('при клике для редактирования устанавливает каретку в нужное место', () => {
    const focus = jest.fn();
    const current = {textContent: 'test', focus};
    jest.spyOn(React, 'createRef').mockReturnValueOnce({current});
    const range = {collapse: jest.fn(), selectNodeContents: jest.fn()};
    const selection = {addRange: jest.fn(), removeAllRanges: jest.fn()};
    window.getSelection = jest.fn().mockReturnValueOnce(selection as unknown as Selection);
    document.createRange = jest.fn().mockReturnValueOnce(range as unknown as Range);
    const wrapper = shallow(<EditableField {...mockProps} hiddenIcon />);

    wrapper.find('span > span').simulate('click')

    expect(focus).toHaveBeenCalled();
    expect(range.selectNodeContents).toHaveBeenCalledWith(current);
    expect(range.collapse).toHaveBeenCalledWith(false);
    expect(selection.removeAllRanges).toHaveBeenCalled();
    expect(selection.addRange).toHaveBeenCalledWith(range);
  });
  it('при клике для редактирования устанавливает каретку в нужное место', () => {
    const focus = jest.fn();
    const current = {textContent: 'test', focus};
    jest.spyOn(React, 'createRef').mockReturnValueOnce({current});
    const range = {collapse: jest.fn(), selectNodeContents: jest.fn()};
    window.getSelection = jest.fn().mockReturnValueOnce(undefined as unknown as Selection);
    document.createRange = jest.fn().mockReturnValueOnce(range as unknown as Range);
    const wrapper = shallow(<EditableField {...mockProps} hiddenIcon />);

    wrapper.find('span > span').simulate('click');

    expect(focus).toHaveBeenCalled();
    expect(range.selectNodeContents).toHaveBeenCalledWith(current);
    expect(range.collapse).toHaveBeenCalledWith(false);
  });
});
