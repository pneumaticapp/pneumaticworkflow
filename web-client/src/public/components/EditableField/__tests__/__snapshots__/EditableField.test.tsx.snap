// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableField корректно рендерится 1`] = `
<span
  className="edit"
  id="test-id"
>
  <span
    className="edit-field"
    contentEditable={false}
    onBlur={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    suppressContentEditableWarning={true}
  >
    test
  </span>
  <i
    className="simple-icon-pencil edit-icon"
    onClick={[Function]}
  />
</span>
`;
